app/controllers/checkout_controller.rb

```diff
@@ -152,8 +154,10 @@ def handle_redirect_from_stripe
 
   def checkout_workflow(shipping_method_id)
     while @order.state != "complete"
-      if @order.state == "payment" && redirect_to_payment_gateway
-        return
+      if @order.state == "payment"
+        return if redirect_to_payment_gateway
+
+        return action_failed unless @order.process_payments!
```

Comment:
This line is taking the deleted `before_transition` callback in the checkout.rb and moving it here, since in the checkout flow we do want to ensure that we process the payment before moving the order to `complete`.